client-csharp/SessionPool.cs

```diff
@@ -0,0 +1,642 @@
+            catch(TException e){
+                client_lst.Add(client);
+                var message = String.Format("counld not get time zone");
+                throw new TException(message, e); 
+            }
+            client_lst.Add(client);
+            return resp.TimeZone;
+        }
+
+        public async Task<Client> create_and_open(bool enableRPCCompression){          
+            TcpClient tcp_client = new TcpClient(this.host, this.port);
+            TSIService.Client client;
+            long sessionId, statementId;
+            var transport = new TFramedTransport(new TSocketTransport(tcp_client, null));
+            if(!transport.IsOpen){
+                try{
+                    await transport.OpenAsync(new CancellationToken());
+                }
+                catch(TTransportException){
+                    throw;
+                }
+            }
+            if(enableRPCCompression){
+                client = new TSIService.Client(new TCompactProtocol(transport));
+            }else{
+                client = new TSIService.Client(new TBinaryProtocol(transport));
+            }
+            var open_req = new TSOpenSessionReq(protocol_version, zoneId);
+            open_req.Username = username;
+            open_req.Password = password;
+            try{
+                var open_resp = await client.openSessionAsync(open_req);
+                if(open_resp.ServerProtocolVersion != protocol_version){
+                    var message = String.Format("Protocol Differ, Client version is {0} but Server version is {1}", protocol_version, open_resp.ServerProtocolVersion);
+                    throw new TException(message, null);
+                }
+                if (open_resp.ServerProtocolVersion == 0){
+                    throw new TException("Protocol not supported", null);
+                }
+                sessionId = open_resp.SessionId;
+                statementId = await client.requestStatementIdAsync(sessionId);
+            }
+            catch(Exception){
+                transport.Close();
+                throw;
+            }
+            is_close = false; 
+            var return_client = new Client();
+            return_client.client = client;
+            return_client.sessionId = sessionId;
+            return_client.statementId = statementId;
+            return_client.transport = transport;
+            return return_client;       
+        }
+        public async Task<int> set_storage_group_async(string group_name){
+            TSStatus status;
+            var client = client_lst.Take();
+            try{
+                status = await client.client.setStorageGroupAsync(client.sessionId, group_name);
+            }
+            catch(TException e){
+                client_lst.Add(client);
+                var err_msg = String.Format("set storage group {0} failed", group_name);
+                throw new TException(err_msg, e);
+            }
+           
+            if(debug_mode){
```

Comment:
Can we extract all such code related to debug to a separated wrapper/decorator class?

Reply:
Cool！